source/godang_ws/src/vision/vision/vision_node.py

```diff
@@ -1,69 +1,118 @@
-from ultralytics import YOLO
-# from vision_class import DistanceCalculator
+from ultralytics import YOLOv10
 import rclpy
 from rclpy.node import Node
 import numpy as np
 import cv2
 
-# class DistanceCalculator:
-#     def __init__(self, focal_length, object_height):
-#         self.focal_length = focal_length  
-#         self.object_height = object_height  # Actual height of the object in meters
+class_names = ['purple', 'red']
 
-#     def calculate_distance(self, object_height_in_image):
-#         distance = (self.focal_length * self.object_height) / object_height_in_image
-#         return distance
+camera_matrix = np.array([[1029.138061543091, 0, 1013.24017],
+                          [0, 992.6178560916601, 548.550898],
+                          [0, 0, 1]])
 
-#     def process_detections(self, results):
+dist_coeffs = np.array([ 0.19576996 ,-0.24765409, -0.00625207 , 0.0039396 ,  0.10282869])
+
+new_camera_matrix = np.array([[1.08832011e+03, 0.00000000e+00 ,1.02215651e+03],
+                                [0.00000000e+00,1.05041880e+03 ,5.39881529e+02],
+                                [0.00000000e+00, 0.00000000e+00, 1.00000000e+00]])
+roi = [0, 0, 1919, 1079]
+
+def detect_objects(frame):
+    model = YOLOv10("../../../../../models/bestv10_redball.pt")
+    results = model(frame)
+    detections = []
+    for bbox in results:
+        boxes = bbox.boxes
+        cls = boxes.cls.tolist()
+        xyxy = boxes.xyxy
+        conf = boxes.conf
+        for i, class_index in enumerate(cls):
+            class_name = class_names[int(class_index)]
+            x1, y1, x2, y2 = map(int, xyxy[i])
+            detections.append([x1, y1, x2, y2])
+    list_of_ball = [detections, conf, class_name]
+    return list_of_ball
+
+def image_to_robot_coordinates(u, v, depth):
+    fx = new_camera_matrix[0, 0]
+    fy = new_camera_matrix[1, 1]
+    cx = new_camera_matrix[0, 2]
+    cy = new_camera_matrix[1, 2]
+
+    x_norm = (u - cx) / fx
+    y_norm = (v - cy) / fy
+    depth_T = depth + 0.215
+    xyz_robot_coordinates = [x_norm, y_norm, depth_T]
+    return xyz_robot_coordinates
+    
+
+def computeBallPosRobotframe(list_of_ball):
+    ## radio of the ball
+    radio_threshold = 1.2
+    ## check if there are any balls
+    if len(list_of_ball) == 0:
+        return None
+    ## check if there is any red balls
+    red_ball = False
+    for i in range(len(list_of_ball)):
+        if list_of_ball[i][2] == 'red':
+            red_ball = True    
+            break
+    
+    ## if there is no red ball
+    if red_ball == False:
+        return 999
+
+    ## first choose most confident ball and match radio
+    sorted_conf_ball = sorted(list_of_ball, key=lambda x: x[1], reverse=True)
+    for i in range(len(sorted_conf_ball)):
+        diff_x = sorted_conf_ball[i][0][2] - sorted_conf_ball[i][0][0]
+        diff_y = sorted_conf_ball[i][0][3] - sorted_conf_ball[i][0][1]
+        if sorted_conf_ball[i][2] == 'red' and diff_x/diff_y < radio_threshold:
```

Comment:
ควรจะกันทั้งสองฝั่งนะ เช่นบอลผอมมากๆ ก็ไม่เอา (สะท้อนแสง)

Reply:
edit to 
def computeBallPosRobotframe(list_of_ball):
    ## radio of the ball
    radio_threshold = 0.08

    ## Tolarance of the wight and high ball
    tolarance = 5

    ## check if there are any balls
    if list_of_ball == []:
        return []
    ## check if there is any red balls
    red_ball = False
    for i in range(len(list_of_ball)):
        if list_of_ball[i][2] == 'red':
            red_ball = True    
            break
    
    ## if there is no red ball
    if not red_ball:
        ball_pos = []
        return ball_pos

    ## first choose most confident ball and match radio
    # print(list_of_ball)
    if list_of_ball != []:
        sorted_conf_ball = sorted(list_of_ball, key=lambda x: x[1], reverse=True)
        for i in range(len(sorted_conf_ball)):
            diff_x = sorted_conf_ball[i][0][2] - sorted_conf_ball[i][0][0]
            diff_y = sorted_conf_ball[i][0][3] - sorted_conf_ball[i][0][1]
            if sorted_conf_ball[i][2] == 'red' and abs(diff_x - diff_y) < tolarance:
                ## compute the center of the ball
                x1, y1, x2, y2 = sorted_conf_ball[i][0]
                u = x1 + (x2 - x1) / 2
                v = y1 + (y2 - y1) / 2

                ## Get depth
                depth_x = (real_diameter * focal_length_x) / (x2-x1)
                

                ## compute the image_to_robot_coordinates
                X, Y, Z = image_to_robot_coordinates(u, v, depth_x)
                # 0.47 is the high from the camera to the ground
                if Y - 0.47 < radio_threshold + 0.19: # 0.19 is the radius of the ball
                    ball_pos = [X, Y, Z]
                    return ball_pos
                else:
                    pass

                
        else:
            ball_pos = []
    return ball_pos